### PR TITLE
Refactored buildings

### PR DESCRIPTION
**Pull request by kohpangwei**
Prettified action names and refactored action routes to support that
Added food, foodPerTurn
Added Iron, Agriculture

@marcgrr @z-chen @katfishi PTAL

### CODE REVIEW

File: tta-server/app/controllers/GameStateResource.scala [hunk 1]
```diff
@@ -4,11 +4,7 @@ import javax.inject.Inject
 import javax.inject.Singleton
 
 import logic.Logic
-import models.ActionId
-import models.Bronze
-import models.GameState
-import models.PlayerIndex
-import models.PlayerState
+import models._
```
Comment (marcgrr) [inline tta-server/app/controllers/GameStateResource.scala]: don't use wildcard imports, for reasons described here https://docs.dkandu.me/practices/languages/scala/imports.html

IntelliJ IDEA > Prefrerences > Code Style > Scala > Imports lets you configure intellij to not automatically do this:
screen shot 2015-01-22 at 9 04 35 am

File: tta-server/app/logic/Logic.scala [hunk 2]
```diff
@@ -1,32 +1,24 @@
...
   def deriveGameState(gameState: GameState): DerivedGameState = {
-    val defaultCandidateActions: immutable.Seq[Action] = List(BuildBronze)
+    val defaultCandidateActionMap = gameState.activePlayerState.researchedBuildings.map (
+      building => {
+        ActionId("build" + building.prettyName) -> Building.generateBuildAction(building)
+      } //why does flatMap have an error here?
+    ).toMap
 
-    val allowedActions = defaultCandidateActions.filter { action =>
+    val allowedActions = defaultCandidateActionMap.filter { case (index, action)=>
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: space between `)` and `=>`

File: tta-server/app/logic/Logic.scala [hunk 3]
```diff
@@ -1,32 +1,24 @@
...
-import models.PlayerState
+import models._
 
 import scala.collection.immutable
 
 object Logic {
   
   def deriveGameState(gameState: GameState): DerivedGameState = {
-    val defaultCandidateActions: immutable.Seq[Action] = List(BuildBronze)
+    val defaultCandidateActionMap = gameState.activePlayerState.researchedBuildings.map (
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]:
```
val defaultCandidateActionMap = gameState.activePlayerState.researchedBuildings.map {
  building =>
    ...
}.toMap
```

File: tta-server/app/logic/Logic.scala [hunk 4]
```diff
@@ -1,32 +1,24 @@
...
+      building => {
+        ActionId("build" + building.prettyName) -> Building.generateBuildAction(building)
+      } //why does flatMap have an error here?
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: `flatMap` expects your function to return a sequence-y thing but this function returns a single element, so `map` is more correct

File: tta-server/app/models/DerivedPlayerState.scala [hunk 5]
```diff
@@ -6,10 +6,14 @@ import play.api.libs.json.OFormat
...
-    orePerTurn: Int) {
+    orePerTurn: Int,
+    foodPerTurn: Int) {
 
   def +(other: DerivedPlayerState): DerivedPlayerState = {
-    DerivedPlayerState(orePerTurn + other.orePerTurn)
+    DerivedPlayerState(
+      orePerTurn + other.orePerTurn,
+      foodPerTurn + other.foodPerTurn
+    )
```
Comment (marcgrr) [inline tta-server/app/models/DerivedPlayerState.scala]: closing paren at end of previous line

File: tta-server/app/logic/Logic.scala [hunk 6]
```diff
@@ -45,7 +37,10 @@ object Logic {
   def updatePlayerStateAtEndOfTurn(
       playerState: PlayerState,
       derivedPlayerState: DerivedPlayerState): PlayerState = {
-    playerState.copy(ore = playerState.ore + derivedPlayerState.orePerTurn)
+    playerState.copy(
+      ore = playerState.ore + derivedPlayerState.orePerTurn,
+      food = playerState.food + derivedPlayerState.foodPerTurn
+    )
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: closing paren at end of previous line

File: tta-server/app/models/Population.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package models
+
+object Population {
+
+  def generateIncreasePopulationAction(playerState: PlayerState): Action = new Action{
```
Comment (marcgrr) [inline tta-server/app/models/Population.scala]: this does not need a `playerState`, does it?

also, space between `Action` and `{`

File: tta-server/app/models/PlayerState.scala [hunk 8]
```diff
@@ -7,8 +7,18 @@ import scala.collection.immutable
...
 
 object PlayerState {
   implicit val format: OFormat[PlayerState] = Json.format[PlayerState]
+  val newPlayerState: PlayerState = PlayerState(
+      buildings = List(Bronze(0)),
+      researchedBuildings = List(Bronze(0), Iron(0), Agriculture(0)),
+      ore = 0,
+      food = 0,
+      population = 1
+  )
```
Comment (marcgrr) [inline tta-server/app/models/PlayerState.scala]: closing paren at end of previous line

File: tta-server/app/models/Population.scala [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    val costToBuild = 2
+
+    override def doIt(gameState: GameState): GameState = {
+      gameState.updatedActivePlayerState ( playerState =>
```
Comment (marcgrr) [inline tta-server/app/models/Population.scala]: your positioning and spacing for this anonymous function is correct and you just have the wrong type of braces. use curly braces for multiline anonymous functions

File: tta-server/app/logic/Logic.scala [hunk 10]
```diff
@@ -1,32 +1,28 @@
...
+      } //why does flatMap have an error here?
+    ).toMap
 
-    val allowedActions = defaultCandidateActions.filter { action =>
+    val increasePopulationCandidateActionMap = Map(ActionId("increasePopulation") -> Population.generateIncreasePopulationAction(gameState.activePlayerState))
+
+    val candidateActionMap = buildCandidateActionMap ++ increasePopulationCandidateActionMap
+
+    val allowedActions = candidateActionMap.filter { case (index, action)=>
       action.isValid(gameState)
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: maybe not to be fixed in this revision, but I do not think that there should be an `isValid` function on actions

instead we should structure the action generation in such a way that only allowable actions get generated in the first place.

the reason is that our action generation is built in such a way that there are three types of actions:
1. those that never get generated (eg a build for an unresearched building)
2. those that get generated but filtered out (eg a build for a researched building that you don't have enough resources to build)
3. those that get generated and not filtered out

this means that filtering is really happening in two different places (steps 1 and 2), which makes it harder to reason about

File: tta-server/app/models/Population.scala (hunk 7 above)
Comment (kohpangwei) [inline tta-server/app/models/Population.scala]: It doesn't need a playerState right now, but in the future the costToBuild will depend on the current population size.

Comment (kohpangwei) [inline tta-server/app/models/Population.scala]: Oh, but you're right, doIt already has access to gameState so the Action itself doesn't need it.

File: tta-server/app/logic/Logic.scala (hunk 10 above)
Comment (kohpangwei) [inline tta-server/app/logic/Logic.scala]: Interesting point. Ok! I pulled it out. What do you think?

File: tta-server/app/controllers/GameStateResource.scala [hunk 11]
```diff
@@ -17,15 +13,17 @@ import play.api.mvc.Controller
...
     Map(
-      PlayerIndex(0) -> PlayerState(
-        buildings = List(Bronze(0)),
-        ore = 0),
-      PlayerIndex(1) -> PlayerState(
-        buildings = List(Bronze(0)),
-        ore = 0)))
+      PlayerIndex(0) -> PlayerState.newPlayerState,
+      PlayerIndex(1) -> PlayerState.newPlayerState
+    )
```
Comment (marcgrr) [inline tta-server/app/controllers/GameStateResource.scala]: wrong line

File: tta-server/app/controllers/GameStateResource.scala [hunk 12]
```diff
@@ -17,15 +13,17 @@ import play.api.mvc.Controller
...
+  )
```
Comment (marcgrr) [inline tta-server/app/controllers/GameStateResource.scala]: wrong line

File: tta-server/app/logic/Logic.scala [hunk 13]
```diff
@@ -35,9 +16,26 @@ object Logic {
   def derivePlayerState(playerIndex: PlayerIndex, gameState: GameState): DerivedPlayerState = {
     val playerState = gameState.playerStates(playerIndex)
 
+    // Generate valid DerivedPlayerStates from civil cards in hand
+    val civilHandDerivedPlayerStates = {
+      for (card <- gameState.activePlayerState.civilHand)
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: map

File: tta-server/app/logic/Logic.scala [hunk 14]
```diff
@@ -35,9 +16,26 @@ object Logic {
...
+        yield card.generatePlayActionDerivedPlayerState(gameState)
+    }.toList
+
+    // Generate valid DerivedPlayerStates from techs that have been researched
+    val techsDerivedPlayerStates = {
+      for (tech <- gameState.activePlayerState.techs)
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: map

File: tta-server/app/logic/Logic.scala [hunk 15]
```diff
@@ -35,9 +16,26 @@ object Logic {
...
+
+    // Generate increase population action, if valid
+    val increasePopulationDerivedPlayerStates = List(Population.generateIncreasePopulationActionDerivedPlayerState(gameState))
+
+    // Calculate resource gains from buildings you've already built
     val buildingEffects = playerState.buildings.map(_.derivePlayerState(gameState))
 
-    val allEffects = buildingEffects
+    // Pool everything together
+    val allEffects = civilHandDerivedPlayerStates ++ techsDerivedPlayerStates ++ increasePopulationDerivedPlayerStates ++ buildingEffects
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: too long

File: tta-server/app/models/Building.scala [hunk 16]
```diff
@@ -6,27 +6,98 @@ import play.api.libs.json.JsSuccess
...
+      if (Building.canBuild(building, gameState)) {
+        val buildAction: Action = new Action {
+          override def doIt(gameState: GameState): DeltaPlayerState = DeltaPlayerState.empty.copy(
+            newBuildings = List(building),
+            population = -1,
+            ore = -building.costToBuild
+          )
+        }
+        Map(ActionId("build" + building.prettyName) -> buildAction)
+      }
```
Comment (marcgrr) [inline tta-server/app/models/Building.scala]: `else {` on this line

File: tta-server/app/models/Building.scala [hunk 17]
```diff
@@ -6,27 +6,98 @@ import play.api.libs.json.JsSuccess
...
+      else {
+        Map.empty
+      }
+    }
+
+    DerivedPlayerState.empty.copy(
+      actions = actions
+    )
```
Comment (marcgrr) [inline tta-server/app/models/Building.scala]: previous line

File: tta-server/app/models/Building.scala [hunk 18]
```diff
@@ -6,27 +6,98 @@ import play.api.libs.json.JsSuccess
...
 object Building {
   implicit val format: Format[Building] = {
-    val writes = Writes { o: Building => JsString("building") }
+    val writes = Writes { o: Building => JsString(o.prettyName) }
     val reads = Reads { json => JsSuccess(Bronze(0) : Building) }
+    //TODO: Make it read something sensible
     Format(reads, writes)
   }
+
+  private def canBuild(building: Building, gameState: GameState) : Boolean = {
```
Comment (marcgrr) [inline tta-server/app/models/Building.scala]: no space between ) and :

File: tta-server/app/models/Building.scala [hunk 19]
```diff
@@ -6,27 +6,98 @@ import play.api.libs.json.JsSuccess
...
+    gameState.activePlayerState.ore >= building.costToBuild &&
+      gameState.activePlayerState.population >= 1
+  }
+
+  def generateResearchedDerivedPlayerState(building: Building, gameState: GameState): DerivedPlayerState = {
```
Comment (marcgrr) [inline tta-server/app/models/Building.scala]: move in to trait

File: tta-server/app/models/DeltaPlayerState.scala [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+  def applyDeltaPlayerState(deltaPlayerState: DeltaPlayerState): (PlayerState => PlayerState) = (playerState: PlayerState) => {
+    PlayerState(
+      buildings = playerState.buildings ++ deltaPlayerState.newBuildings,
+      techs = playerState.techs ++ deltaPlayerState.newTechs,
+      civilHand = (playerState.civilHand ++ deltaPlayerState.newCivilHand) diff deltaPlayerState.removedCivilHand,
+      population = playerState.population + deltaPlayerState.population,
+      ore = playerState.ore + deltaPlayerState.ore,
+      food = playerState.food + deltaPlayerState.food,
+      science = playerState.science + deltaPlayerState.science
+    )
```
Comment (marcgrr) [inline tta-server/app/models/DeltaPlayerState.scala]: previous line

File: tta-server/app/models/DeltaPlayerState.scala [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+  val empty = DeltaPlayerState(
+    newBuildings = List.empty,
+    newTechs = List.empty,
+    newCivilHand = List.empty,
+    removedCivilHand = List.empty,
+    population = 0,
+    ore = 0,
+    food = 0,
+    science = 0
+  )
```
Comment (marcgrr) [inline tta-server/app/models/DeltaPlayerState.scala]: previous line

File: tta-server/app/models/DeltaPlayerState.scala [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+  removedCivilHand: immutable.Seq[Card],
+  population: Int,
+  ore: Int,
+  food: Int,
+  science: Int
+)
+
+object DeltaPlayerState {
+
+  def applyDeltaPlayerState(deltaPlayerState: DeltaPlayerState): (PlayerState => PlayerState) = (playerState: PlayerState) => {
```
Comment (marcgrr) [inline tta-server/app/models/DeltaPlayerState.scala]: put `{` before the argument and rocket

File: tta-server/app/models/DeltaPlayerState.scala [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+case class DeltaPlayerState(
+  newBuildings: immutable.Seq[Building],
+  newTechs: immutable.Seq[Tech],
+  newCivilHand: immutable.Seq[Card],
+  removedCivilHand: immutable.Seq[Card],
+  population: Int,
+  ore: Int,
+  food: Int,
+  science: Int
+)
```
Comment (marcgrr) [inline tta-server/app/models/DeltaPlayerState.scala]: previous line

File: tta-server/app/models/DeltaPlayerState.scala [hunk 24]
```diff
@@ -0,0 +1,42 @@
+package models
+
+import models.PlayerState
+
+import collection.immutable
+
+case class DeltaPlayerState(
+  newBuildings: immutable.Seq[Building],
```
Comment (marcgrr) [inline tta-server/app/models/DeltaPlayerState.scala]: 4-space indent for argument lists

File: tta-server/app/models/PlayerState.scala [hunk 25]
```diff
@@ -7,8 +7,21 @@ import scala.collection.immutable
...
+    civilHand: immutable.Seq[Card],
+    population: Int,
+    ore: Int,
+    food: Int,
+    science: Int)
 
 object PlayerState {
   implicit val format: OFormat[PlayerState] = Json.format[PlayerState]
+  val newPlayerState: PlayerState = PlayerState(
+      buildings = List(Bronze(0), Bronze(0), Agriculture(0), Agriculture(0), Philosophy(0)),
```
Comment (marcgrr) [inline tta-server/app/models/PlayerState.scala]: only 2-space indent here

File: tta-server/app/models/Population.scala [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+
+  def generateIncreasePopulationActionDerivedPlayerState(gameState: GameState): DerivedPlayerState = {
+
+    val actions: Map[ActionId, Action] = {
+      if (Population.canIncrease(gameState)) {
+        val increasePopulationAction: Action = new Action {
+          override def doIt(gameState: GameState): DeltaPlayerState = DeltaPlayerState.empty.copy(
+            population = 1,
+            food = -costToIncrease(gameState)
+          )
```
Comment (marcgrr) [inline tta-server/app/models/Population.scala]: previous line

File: tta-server/app/models/Population.scala [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+        }
+        Map(ActionId("increasePopulation") -> increasePopulationAction)
+      }
```
Comment (marcgrr) [inline tta-server/app/models/Population.scala]: `else {` on same line

File: tta-server/app/models/Tech.scala [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+    //TODO: Make it read something sensible
+    Format(reads, writes)
+  }
+
+  private def canResearch(tech: Tech, gameState: GameState) : Boolean = {
+    gameState.activePlayerState.science >= tech.costToResearch &&
+      !gameState.activePlayerState.techs.contains(tech)
+  }
+
+  def generatePlayActionDerivedPlayerState(tech: Tech, gameState: GameState): DerivedPlayerState = {
```
Comment (marcgrr) [inline tta-server/app/models/Tech.scala]: move to trait

File: tta-server/app/models/Tech.scala [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+
+object Tech {
+  implicit val format: Format[Tech] = {
+    val writes = Writes { o: Tech => JsString(o.prettyName) }
+    val reads = Reads { json => JsSuccess(Bronze(0) : Tech) }
+    //TODO: Make it read something sensible
+    Format(reads, writes)
+  }
+
+  private def canResearch(tech: Tech, gameState: GameState) : Boolean = {
```
Comment (marcgrr) [inline tta-server/app/models/Tech.scala]: no space between ) and :

File: tta-server/app/models/Tech.scala [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+      if (Tech.canResearch(tech, gameState)) {
+        val buildAction: Action = new Action {
+          override def doIt(gameState: GameState): DeltaPlayerState = DeltaPlayerState.empty.copy(
+            newTechs = List(tech),
+            science = -tech.costToResearch,
+            removedCivilHand = List(tech)
+          )
+        }
+        Map(ActionId("research" + tech.prettyName) -> buildAction)
+      }
```
Comment (marcgrr) [inline tta-server/app/models/Tech.scala]: `else {` on this line

File: tta-server/app/models/Tech.scala [hunk 31]
```diff
@@ -0,0 +1,52 @@
...
+
+  def generatePlayActionDerivedPlayerState(tech: Tech, gameState: GameState): DerivedPlayerState = {
+    val actions: Map[ActionId, Action] = {
+      if (Tech.canResearch(tech, gameState)) {
+        val buildAction: Action = new Action {
+          override def doIt(gameState: GameState): DeltaPlayerState = DeltaPlayerState.empty.copy(
+            newTechs = List(tech),
+            science = -tech.costToResearch,
+            removedCivilHand = List(tech)
+          )
```
Comment (marcgrr) [inline tta-server/app/models/Tech.scala]: previous line

File: tta-server/app/models/Tech.scala [hunk 32]
```diff
@@ -0,0 +1,52 @@
...
+        Map(ActionId("research" + tech.prettyName) -> buildAction)
+      }
+      else {
+        Map.empty
+      }
+    }
+
+    DerivedPlayerState.empty.copy(
+      actions = actions
+    )
```
Comment (marcgrr) [inline tta-server/app/models/Tech.scala]: previous line

File: tta-server/app/models/Population.scala [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+      else {
+        Map.empty
+      }
+    }
+
+    DerivedPlayerState.empty.copy(
+      actions = actions
+    )
```
Comment (marcgrr) [inline tta-server/app/models/Population.scala]: previous line

**Comment by kohpangwei** [comment]
Style changes made, PTAL @marcgrr

File: tta-server/app/logic/Logic.scala [hunk 34]
```diff
@@ -45,7 +43,10 @@ object Logic {
   def updatePlayerStateAtEndOfTurn(
       playerState: PlayerState,
       derivedPlayerState: DerivedPlayerState): PlayerState = {
-    playerState.copy(ore = playerState.ore + derivedPlayerState.orePerTurn)
+        playerState.copy(
```
Comment (marcgrr) [inline tta-server/app/logic/Logic.scala]: only indent this 2 spaces even though the argument list is 4

File: tta-server/app/models/Building.scala [hunk 35]
```diff
@@ -6,14 +6,40 @@ import play.api.libs.json.JsSuccess
...
   def derivePlayerState(gameState: GameState): DerivedPlayerState
+  def costToBuild: Int
+
+  private def canBuild(gameState: GameState): Boolean = {
+    gameState.activePlayerState.ore >= this.costToBuild &&
+      gameState.activePlayerState.population >= 1
+  }
+
+  override def generateResearchedDerivedPlayerState(gameState: GameState): DerivedPlayerState = {
+    def building = this
```
Comment (marcgrr) [inline tta-server/app/models/Building.scala]: val

File: tta-server/app/models/Tech.scala [hunk 36]
```diff
@@ -0,0 +1,47 @@
...
+
+  private def canResearch(gameState: GameState): Boolean = {
+    gameState.activePlayerState.science >= this.costToResearch &&
+      !gameState.activePlayerState.techs.contains(this)
+  }
+
+  def generateResearchedDerivedPlayerState(gameState: GameState): DerivedPlayerState
+
+  override def generatePlayActionDerivedPlayerState(gameState:GameState): DerivedPlayerState = {
+    def tech = this
```
Comment (marcgrr) [inline tta-server/app/models/Tech.scala]: val

File: tta-server/app/controllers/GameStateResource.scala (hunk 1 above)
Comment (marcgrr) [inline tta-server/app/controllers/GameStateResource.scala]: this is still a wildcard

**Comment by marcgrr** [comment]
nearly there...!!

**Comment by marcgrr** [comment]
LGTM

**Comment by kohpangwei** [comment]
woohoo!

**Pull request merged**